crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,849 @@
+//!
+//! * `code` is an integer representing the kind of an error that happened. Error codes documented
+//!   in the [`error_codes`] module.
+//!
+//! * `message` is a short, one-sentence human-readable description of the error.
+//!
+//! * `data` is an optional field of arbitrary type containing additional information about the error.
+//!
+//! ## Built-in methods
+//!
+//! The Bevy Remote Protocol includes a number of built-in methods for accessing and modifying data
+//! in the ECS. Each of these methods uses the `bevy/` prefix, which is a namespace reserved for
+//! BRP built-in methods.
+//!
+//! ### bevy/get
+//!
+//! Retrieve the values of one or more components from an entity.
+//!
+//! `params`:
+//! - `entity`: The ID of the entity whose components will be fetched.
+//! - `components`: An array of fully-qualified type names of components to fetch.
+//!
+//! `result`: A map associating each type name to its value on the requested entity.
+//!
+//! ### bevy/query
+//!
+//! Perform a query over components in the ECS, returning all matching entities and their associated
+//! component values.
+//!
+//! All of the arrays that comprise this request are optional, and when they are not provided, they
+//! will be treated as if they were empty.
+//!
+//! `params`:
+//! - `data`:
+//!   - `components` (optional): An array of fully-qualified type names of components to fetch.
+//!   - `option` (optional): An array of fully-qualified type names of components to fetch optionally.
+//!   - `has` (optional): An array of fully-qualified type names of components whose presence will be
+//!      reported as boolean values.
+//! - `filter` (optional):
+//!   - `with` (optional): An array of fully-qualified type names of components that must be present
+//!     on entities in order for them to be included in results.
+//!   - `without` (optional): An array of fully-qualified type names of components that must *not* be
+//!     present on entities in order for them to be included in results.
+//!
+//! `result`: An array, each of which is an object containing:
+//! - `entity`: The ID of a query-matching entity.
+//! - `components`: A map associating each type name from `components`/`option` to its value on the matching
+//!   entity if the component is present.
+//! - `has`: A map associating each type name from `has` to a boolean value indicating whether or not the
+//!   entity has that component. If `has` was empty or omitted, this key will be omitted in the response.
+//!
+//! ### bevy/spawn
+//!
+//! Create a new entity with the provided components and return the resulting entity ID.
+//!
+//! `params`:
+//! - `components`: A map associating each component's fully-qualified type name with its value.
+//!
+//! `result`:
+//! - `entity`: The ID of the newly spawned entity.
+//!
+//! ### bevy/destroy
+//!
+//! Despawn the entity with the given ID.
+//!
+//! `params`:
+//! - `entity`: The ID of the entity to be despawned.
+//!
+//! `result`: null.
```

Comment:
If there is a `result` at all, that means that the request succeeded. If there is an `error`, it failed. It's basically like `Result<T, E>` in that way, if you want to think of it like that.

Reply:
Right but I’m thinking of someone writing something like a JavaScript application to interface with this. They would have to use `response.hasOwnProperty("result")`, `response.error === undefined`, or `response.result !== undefined`. And it would be an easy mistake to use `response.error !== undefined` or `response.result !== null`, since those will only return `true` on success for any non-null-returning request.

I guess this shouldn't be an issue if they read the documentation and understand what kinds of responses to expect, but something to consider.

Reply:
I see what you mean, but this is also just how success and error are defined in response objects according to the JSON-RPC spec, so they would have to do that in order to handle responses correctly in any case.

Reply:
Ah okay that makes sense it's part of the spec. Thanks for the clarification!